File: packages/javascript/src/gradient-noise-2d-01.js
import { cornerHash } from "./internal/noise-hash.js";
// lowbias32 mixer: https://github.com/skeeto/hash-prospector (Unlicense).
// See THIRD_PARTY_NOTICES.md; lattice composition and gradient ordering are project choices.
const MIN_QUERY_COORDINATE = -9_007_199_254_740_991;
const MAX_QUERY_COORDINATE = 9_007_199_254_740_991;
const hasOwn = (value, key) => Object.prototype.hasOwnProperty.call(value, key);

/** Error with one of the stable gradient-noise-2d-01 contract codes. */
export class GradientNoise2D01Error extends Error {
  constructor(code) {
    super(code);
    this.name = "GradientNoise2D01Error";
    this.code = code;
  }
}

function fail(code) {
  throw new GradientNoise2D01Error(code);
}

function normalizedZero(value) {
  return value === 0 ? 0 : value;
}

function finiteQueryCoordinate(value) {
  return typeof value === "number" && Number.isFinite(value) &&
    value >= MIN_QUERY_COORDINATE && value < MAX_QUERY_COORDINATE;
}

function dataProperty(object, key) {
  const descriptor = Object.getOwnPropertyDescriptor(object, key);
  return descriptor !== undefined && "value" in descriptor ? descriptor.value : undefined;
}

function validateSeed(params) {
  if (params === null || typeof params !== "object" || Array.isArray(params)) {
    fail("INVALID_INPUT");
  }
  const keys = Reflect.ownKeys(params);
  if (keys.length !== 1 || !hasOwn(params, "seed")) {
    fail("INVALID_INPUT");
  }
  const seed = dataProperty(params, "seed");
  if (typeof seed !== "number" || !Number.isFinite(seed) || !Number.isInteger(seed) ||
      seed < 0 || seed > 0xffff_ffff) {
    fail("INVALID_INPUT");
  }
  return normalizedZero(seed);
}

function validateTuple(query) {
  if (!Array.isArray(query) || query.length !== 2 || !hasOwn(query, 0) || !hasOwn(query, 1)) {
    fail("INVALID_QUERY");
  }
  const keys = Reflect.ownKeys(query);
  if (keys.length !== 3 || !keys.includes("length") || !keys.includes("0") || !keys.includes("1")) {
    fail("INVALID_QUERY");
  }
  const x = dataProperty(query, "0");
  const y = dataProperty(query, "1");
  if (!finiteQueryCoordinate(x) || !finiteQueryCoordinate(y)) {
    fail("INVALID_QUERY");
  }
  return [x, y];
}

function validateScalars(x, y) {
  if (!finiteQueryCoordinate(x) || !finiteQueryCoordinate(y)) {
    fail("INVALID_QUERY");
  }
}

function gradientDot(hash, dx, dy) {
  let gx;
  let gy;
  switch (hash & 7) {
    case 0: gx = 1; gy = 0; break;
    case 1: gx = -1; gy = 0; break;
    case 2: gx = 0; gy = 1; break;
    case 3: gx = 0; gy = -1; break;
    case 4: gx = 1; gy = 1; break;
    case 5: gx = -1; gy = 1; break;
    case 6: gx = 1; gy = -1; break;
    default: gx = -1; gy = -1; break;
  }
  const px = gx * dx;
  const py = gy * dy;
  return px + py;
}

function fade(t) {
  const a = t * t;
  const b = a * t;
  const c = 6 * t;
  const d = c - 15;
  const e = t * d;
  const f = e + 10;
  return b * f;
}

function lerp(a, b, t) {
  const difference = b - a;
  const product = t * difference;
  return a + product;
}

/**
 * Create an immutable single-octave 2D portable gradient-noise field.
 *
 * Motivating sketch: 2018/Generativos/pelines. No artistic default or encouraged
 * parameter range is established; seed is explicit and coordinates use lattice units.
 * The field is pure and seed-only. It is independently specified for CP1 rather
 * than a compatibility wrapper around a host noise implementation.
 */
export function gradientNoise2D01(params) {
  const seed = validateSeed(params);

  function sampleScalars(x, y) {
    const i = Math.floor(x);
    const j = Math.floor(y);
    const u = x - i;
    const v = y - j;
    const n00 = gradientDot(cornerHash(seed, i, j), u, v);
    const n10 = gradientDot(cornerHash(seed, i + 1, j), u - 1, v);
    const n01 = gradientDot(cornerHash(seed, i, j + 1), u, v - 1);
    const n11 = gradientDot(cornerHash(seed, i + 1, j + 1), u - 1, v - 1);
    const fx = fade(u);
    const fy = fade(v);
    const bottom = lerp(n00, n10, fx);
    const top = lerp(n01, n11, fx);
    const raw = lerp(bottom, top, fy);
    const scaled = 0.5 * raw;
    let result = 0.5 + scaled;
    if (result < 0) result = 0;
    if (result > 1) result = 1;
    return normalizedZero(result);
  }

  function sample(x, y) {
    if (arguments.length === 1) {
      const query = validateTuple(x);
      return sampleScalars(query[0], query[1]);
    }
    if (arguments.length === 2) {
      validateScalars(x, y);
      return sampleScalars(x, y);
    }
    throw new TypeError("sample expects either [x, y] or x, y");
  }

  function serialize() {
    return { seed };
  }

  const field = { sample, serialize };
  Object.defineProperty(field, "toJSON", { value: serialize });
  return Object.freeze(field);
}

File: packages/java/src/main/java/org/procedurals/fields/GradientNoise2D01.java
package org.procedurals.fields;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Pure single-octave scalar field, motivated by 2018/Generativos/pelines.
 * Independently specified algorithm; not Processing noise compatibility.
 * Seed is explicit unsigned32; no artistic default or encouraged parameter range
 * is established. Coordinates require exact-safe lower and upper lattice corners.
 * Contract: field.gradient-noise-2d-01 0.1.0. lowbias32 mixer provenance:
 * https://github.com/skeeto/hash-prospector (Unlicense); see THIRD_PARTY_NOTICES.md.
 */
public strictfp final class GradientNoise2D01 {
    public static final class NoiseException extends IllegalArgumentException {
        public final String code;
        public NoiseException(String code) { super(code); this.code=code; }
    }
    private static final int[][] GRAD={{1,0},{-1,0},{0,1},{0,-1},{1,1},{-1,1},{1,-1},{-1,-1}};
    private final int seed;
    private GradientNoise2D01(int seed) { this.seed=seed; }

    public static GradientNoise2D01 create(Object input) {
        if(!(input instanceof Map)) { fail("INVALID_INPUT"); return null; }
        Map<?,?> params=(Map<?,?>)input;
        if(params.size()!=1 || !params.containsKey("seed")) fail("INVALID_INPUT");
        double n=number(params.get("seed"),"INVALID_INPUT");
        if(n<0 || n>4294967295L || n!=Math.floor(n)) fail("INVALID_INPUT");
        return new GradientNoise2D01((int)(long)n);
    }
    public Map<String,Object> serialize() {
        Map<String,Object> result=new LinkedHashMap<String,Object>();
        result.put("seed",((long)seed)&0xffffffffL); return result;
    }
    /** Interchange tuple route; the two-scalar overload avoids query-array allocation. */
    public double sample(Object point) {
        if(!(point instanceof List) || ((List<?>)point).size()!=2) { fail("INVALID_QUERY"); return 0; }
        List<?> p=(List<?>)point; return sample(p.get(0),p.get(1));
    }
    public double sample(Object x,Object y) {
        double a=number(x,"INVALID_QUERY"); validate(a);
        double b=number(y,"INVALID_QUERY"); return sample(a,b);
    }
    /** No state consumption or query allocation; exact binary64 result in [0,1]. */
    public double sample(double x,double y) {
        validate(x); validate(y);
        long i=(long)Math.floor(x),j=(long)Math.floor(y);
        double u=x-(double)i,v=y-(double)j;
        double um=u-1,vm=v-1;
        double n00=dot(seed,i,j,u,v),n10=dot(seed,i+1,j,um,v);
        double n01=dot(seed,i,j+1,u,vm),n11=dot(seed,i+1,j+1,um,vm);
        double fx=fade(u),fy=fade(v);
        double bottom=lerp(n00,n10,fx),top=lerp(n01,n11,fx);
        double raw=lerp(bottom,top,fy),scaled=.5*raw,result=.5+scaled;
        if(result<=0) return 0.0;
        return result>=1 ? 1.0 : result;
    }
    static int mix(int v) {
        v^=v>>>16; v*=0x7feb352d; v^=v>>>15; v*=0x846ca68b; return v^(v>>>16);
    }
    static int corner(int seed,long i,long j) { return mix(mix(seed^(int)i^0x9e3779b9)^(int)j^0x85ebca6b); }
    private static double dot(int seed,long i,long j,double x,double y) {
        int[] g=GRAD[corner(seed,i,j)&7]; double px=g[0]*x,py=g[1]*y; return px+py;
    }
    private static double fade(double t) { double a=t*t,b=a*t,c=6*t,d=c-15,e=t*d,f=e+10; return b*f; }
    private static double lerp(double a,double b,double t) { double d=b-a,p=t*d; return a+p; }
    private static void validate(double q) {
        if(Double.isNaN(q) || Double.isInfinite(q) || q < -9007199254740991L || q>=9007199254740991L) fail("INVALID_QUERY");
    }
    private static double number(Object value,String error) {
        if(!(value instanceof Byte || value instanceof Short || value instanceof Integer || value instanceof Long || value instanceof Float || value instanceof Double)) { fail(error); return 0; }
        double n=((Number)value).doubleValue();
        if(Double.isNaN(n)||Double.isInfinite(n)) fail(error);
        return n;
    }
    private static void fail(String code) { throw new NoiseException(code); }
}
